Add sample: Extended Dynamic State2

Mateusz-Kowalewski-Mobica opened the pull request:
## Description

This PR adds a sample for presenting possibilities to utilize features included in the extension VK_EXT_extended_dynamic_state2. For creating conditions to properly show the capabilities of features, a new scene was proposed (merged already). One of five features was excluded from the sample because of creating complexity and reducing transparency.

![Sample_image2](https://user-images.githubusercontent.com/113357738/212025291-e64bef03-da78-4859-b4ef-4f0bd3ce906a.png)

## Features

### Depth Bias Enable
For presenting the possibility of using a depth bias feature, a z-fighting phenomenon was created (two cubes with 4 faces overlapping). After applying depth bias to one model, the negative effect disappears.
### Primitive Restart Enable
To use primitive restart feature, one model data was created in runtime with indices using triangle strip topology (with "UINT32_MAX" as restart value)
### Rasterizer Discard Enable
Rasterizer discarding can be applied to a few models in the scene. Visually objects disappear, because of skipping the rasterization process for them.
### Patch Control Points
Patch control points value can be changed in runtime for one model (wireframe).

## Notes

Sample was tested on next 2 configurations:
Linux Ubuntu 20.04
Nvidia Quadro P600
Intel i7-8750H
32 GB RAM
and
Windows 10
Nvidia Quadro P620
Intel i7-11800H
32 GB RAM

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

Mateusz-Kowalewski-Mobica committed:
Add sample: Extended Dynamic State2

Mateusz-Kowalewski-Mobica committed:
Adjust code for new tessellation object

Mateusz-Kowalewski-Mobica committed:
Background shader implemented

Mateusz-Kowalewski-Mobica committed:
Clang Tidy check fixing

Mateusz-Kowalewski-Mobica committed:
Doxygen documentation added

Mateusz-Kowalewski-Mobica committed:
Fix warnings in other OS platforms

Mateusz-Kowalewski-Mobica committed:
Clang fix

Seweryn-Zielas-Mobica committed:
Create README for Sample

* Create README.md

Create README prototype for extended dynamic state 2

* Update samples README to include dynamic state 2

* Fix typo in samples README form previous commit

* Update README.md

* Remove obsolete code from draw calls code snippet

* Update main extensions-sample README.md

* Delete "dramatically" from extended dynamic state2

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/KhronosGroup/Vulkan-Samples?pullRequest=592) <br/>All committers have signed the CLA.

Mateusz-Kowalewski-Mobica committed:
Copyright/Clang fixes

Mateusz-Kowalewski-Mobica committed:
CMakeLists.txt Author change

Mateusz-Kowalewski-Mobica commented:
@tomadamatkinson Hello Tom,
This is the second sample that we want to propose to merge. I can't add reviewers. Can you help me with that? :)

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,443 @@
+
+## Overview
+
+This sample demostrates hot to use `VK_EXT_extended_dynamic_state2` extension, which eliminates the need to create multiple pipeline in case of specific different parameters.
```
typos:
    hot -> how
    pipeline -> pipelines

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,443 @@
+
+This sample demostrates hot to use `VK_EXT_extended_dynamic_state2` extension, which eliminates the need to create multiple pipeline in case of specific different parameters.
+
+This extension changes how `Depth Bias`, `Primitive Restart`, `Rasterizer Discard` and `Patch Control Points` are managed. Instead of static description during pipeline creation, this extension allows developers to change those parameters by using function before every draw.
```
missing 'a' between 'using' and 'function'

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
````diff
@@ -0,0 +1,443 @@
+Previously developers had to create multiple pipelines for different parameters in `Depth Bias`, `Primitive Restart`, `Rasterizer Discard` and `Patch Control Points`. This is illustrated in static/non-dynamic pipeline creation.
+
+```C++
+...
````
Different indentations in this snippet.

Karina-Murawko-Wisniewska-Mobica replied:
fixed in all code snippets

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
````diff
@@ -0,0 +1,443 @@
+...
+```
+
+In approach above, if we would want to change patch control points number, then for each different number we would need to create new pipeline.
````
Missing 'the' between 'change' and 'patch';
Missing 'a' between 'create' and 'new'

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach replied:
"In approach above" -> "In the above approach" ?

Karina-Murawko-Wisniewska-Mobica replied:
updated

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
````diff
@@ -0,0 +1,443 @@
+...
+```
+
+And now, thanks to dynamic state 2, we can change parameters before each corresponding draw call. 
````
Maybe, instead of 'dynamic state 2': 'VK_EXT_extended_dynamic_state2'?

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,443 @@
+
+In approach above, if we would want to change patch control points number, then for each different number we would need to create new pipeline.
+
+However, with dynamic state 2 the number of pipelines can be reduced because of possibility to change parameters of `Depth Bias`, `Primitive Restart`, `Rasterizer Discard` and `Patch Control Points` by calling `cmdSetDepthBiasEnableEXT`, `cmdSetPrimitiveRestartEnableEXT`, `cmdSetRasterizerDiscardEnableEXT` and `cmdSetPatchControlPointsEXT` before `draw_model`.
```
Maybe: 'because of' -> 'by the';
The four function names should be complete here, that is start with 'vkCmd';
Maybe add a 'respectively' after the last function name;
Maybe 'before drawing the model' or 'before calling `draw_model`', instead of 'before `draw_model`'.

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,443 @@
+
+However, with dynamic state 2 the number of pipelines can be reduced because of possibility to change parameters of `Depth Bias`, `Primitive Restart`, `Rasterizer Discard` and `Patch Control Points` by calling `cmdSetDepthBiasEnableEXT`, `cmdSetPrimitiveRestartEnableEXT`, `cmdSetRasterizerDiscardEnableEXT` and `cmdSetPatchControlPointsEXT` before `draw_model`.
+
+With usage of above functions we can reduce number of pipelines.
```
Missing 'the' between 'With' and 'usage'.

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
````diff
@@ -0,0 +1,443 @@
+
+```C+
+...
+    VkPipelineInputAssemblyStateCreateInfo input_assembly_state =
````
The actual changes from the previous code snippet are difficult to see due to the sheer length of the snippets.
You could add some eye-catching comments to highlight the actual difference.

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
````diff
@@ -0,0 +1,443 @@
+
+```
+
+The usage of depth bias dynamic state is implemented in `draw_from_scene` function which is described below.
````
Missing 'the' between 'in' and '`draw_from_scene`'

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,443 @@
+	std::vector<VkDynamicState> dynamic_state_enables = {
+	    VK_DYNAMIC_STATE_VIEWPORT,
+	    VK_DYNAMIC_STATE_SCISSOR,
+	    VK_DYNAMIC_STATE_PRIMITIVE_TOPOLOGY_EXT,
```
Dynamic usage of primitive topology isn't mentioned anywhere in this readme.

MarcinKlosowskiMobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,443 @@
+
+
+## Enabling the Extension
+The extended dynamic state 2 api is provided in Vulkan 1.3 and the appropriate headers / SDK is required.
```
I think that's not quite right.
VK_EXT_extended_dynamic_state2 requires Vulkan 1.0 and VK_KHR_get_physical_device_properties2. It is (partly?) promoted to Vulkan 1.3.

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+ */
+void ExtendedDynamicState2::load_assets()
+{
+	vkb::GLTFLoader loader{get_device()};
```
Just call load_scene here, from VulkanSample.
No need to have your own scene member in this sample, then.

Szymon-Rajczakowski replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+	// Store all scene nodes in a linear vector for easier access
+	for (auto &mesh : scene->get_components<vkb::sg::Mesh>())
+	{
+		for (auto &node : mesh->get_nodes())
```
Is this loop nesting correct?
Get each submesh again for each node?
In the scene loaded here, there are is just one node and one submesh, but is that in general the case?

Zbigniew-Jedlinski-Mobica replied:
This part is written for more complex solution where one mesh can have many nodes and one node can have many submesh
The loops make code more cleary and readable.

Karina-Murawko-Wisniewska-Mobica replied:
@asuessenbach 
Is it ok to leave the nested loop?

asuessenbach replied:
I see... I would have expected something like traversing the scene tree, but this is gets you the correct stuff as well. Maybe in a different order, but that shouldn't matter here.

Karina-Murawko-Wisniewska-Mobica replied:
The fix is proposed on the fork, currently under review: https://github.com/Mateusz-Kowalewski-Mobica/Vulkan-Samples/pull/16

Zbigniew-Jedlinski-Mobica replied:
@asuessenbach
Above is a proposition of add the meshes found from the scene tree

asuessenbach replied:
Looks good!
But note, that I already considered this issue as resolved.

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+{
+	/* Baseline uniform buffer */
+	ubo_baseline.projection = camera.matrices.perspective;
+	ubo_baseline.view       = camera.matrices.view * glm::mat4(1.f);
```
Why do you multiply the view matrix with an identity matrix here?

MarcinKlosowskiMobica replied:
The multiplication is redundant so it has been removed.

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+
+	/* Tessellation uniform buffer */
+	ubo_tess.projection          = camera.matrices.perspective;
+	ubo_tess.modelview           = camera.matrices.view * glm::mat4(1.0f);
```
Again: multiplication with an identity matrix?

MarcinKlosowskiMobica replied:
The multiplication is redundant so it has been removed.

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+
+	/* Background uniform buffer */
+	ubo_background.projection           = camera.matrices.perspective;
+	ubo_background.background_modelview = camera.matrices.view;
```
The very same data has three different names in the three UBOs: view, modelview, and background_modelview. Why's that?

Karina-Murawko-Wisniewska-Mobica replied:
Changed with UBO refactoring.

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+	/* Note: Using Reversed depth-buffer for increased precision, so Greater depth values are kept */
+	VkPipelineDepthStencilStateCreateInfo depth_stencil_state =
+	    vkb::initializers::pipeline_depth_stencil_state_create_info(
+	        VK_TRUE, /* changed */
```
What does this 'changed' mean here?

Szymon-Rajczakowski replied:
fixed, '/* changed */' has been removed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+		range.baseArrayLayer = 0;
+		range.layerCount     = VK_REMAINING_ARRAY_LAYERS;
+
+		VkImageSubresourceRange depth_range{range};
```
depth_range and range are never used.

Zbigniew-Jedlinski-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+		{
+			update_uniform_buffers();
+		}
+		if (drawer.input_float("Tessellation Factor", &gui_settings.tess_factor, 1.0f, 1))
```
Maybe use a slider_float here with a min value of 1.0f here?

Karina-Murawko-Wisniewska-Mobica replied:
changed to slider_float

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+			update_uniform_buffers();
+		}
+
+		if (drawer.input_float("Patch Control Points", &gui_settings.patch_control_points_float, 1.0f, 1))
```
You could directly use slider_int(), or introduce some input_int() to the Drawer.
Besides that: is changing the patch control points by UI meaningfull at all? Shouldn't that be controlled by the geometries?

Karina-Murawko-Wisniewska-Mobica replied:
yes you are right, I'll remove it from GUI and try to come with other idea of presenting dynamic control points change in pipeline 

we are planning to add second model using quads (e.g. plane) to show dynamic change of control points

Karina-Murawko-Wisniewska-Mobica replied:
changing patch_control_points removed from GUI

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+			{
+				gui_settings.patch_control_points_float = 1;
+			}
+			gui_settings.patch_control_points = static_cast<int>(roundf(gui_settings.patch_control_points_float));
```
No update_uniform_buffers() here?

Karina-Murawko-Wisniewska-Mobica replied:
changing patch_control_points removed from GUI, change of patch control points will be used with new model

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+	{
+		const char *col_names = {"Name"};
+		if (drawer.checkbox("Selection effect active", &gui_settings.selection_active))
+		{
```
Should something be done, here?

Zbigniew-Jedlinski-Mobica replied:
no if-statement was removed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+		{
+		}
+		ImGui::Columns(2, "Name");
+		ImGui::SetColumnWidth(0, 150);
```
150 is a bit too small for the object names.
Why are you using two columns here at all?
Wouldn't a combo_box be a better choice for the object names?

Zbigniew-Jedlinski-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+			break;
+		}
+	}
+	return i;
```
Maybe instead:
```
 	return std::distance(scene_node->begin(),
	                     std::find_if(scene_node->begin(), scene_node->end(), [&name](SceneNode const &node) { return node.node->get_name() == name; }));

```

MarcinKlosowskiMobica replied:
corrected as per the review comment

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+			gui_settings.time_tick = false;
+		}
+		new_mat->base_color_factor.w += accumulated_diff;
+	}
```
Maybe instead:
```
	if ( gui_settings.time_tick)
	{
		accumulated_diff += rise ? alpha_step : -alpha_step;
		gui_settings.time_tick = false;
	}
	new_mat->base_color_factor.w += accumulated_diff;

```

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+{
+	static bool                        rise              = false;
+	static int                         previous_obj_id   = gui_settings.selected_obj;
+	static const vkb::sg::PBRMaterial *previous_material = original_mat;
```
previous_material is never used.

Szymon-Rajczakowski replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+ * @fn void ExtendedDynamicState2::selection_indicator(const vkb::sg::PBRMaterial *original_mat, vkb::sg::PBRMaterial *new_mat)
+ * @brief Changing alpha value to create blinking effect on selected model
+ */
+void ExtendedDynamicState2::selection_indicator(const vkb::sg::PBRMaterial *original_mat, vkb::sg::PBRMaterial *new_mat)
```
The only thing that's used from this helper function is new_mat->base_color_factor.
So why not just return that and completely skip that new_mat?

Szymon-Rajczakowski replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+		/* Pass data for the current node via push commands */
+		auto node_material            = dynamic_cast<const vkb::sg::PBRMaterial *>(node[i].sub_mesh->get_material());
+		push_const_block.model_matrix = node[i].node->get_transform().get_world_matrix();
+		if (i != gui_settings.selected_obj ||
```
Maybe change the logic here:
`		if (gui_settings.selection_active && ( i == gui_settings.selected_obj))
`
I think, that's easier to grasp.

MarcinKlosowskiMobica replied:
modified as per the review comment

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+	constexpr float move_step  = 0.0005;
+	static float    time_pass  = 0;
+	time_pass += delta_time;
+	static glm::vec3 translation = scene_nodes[SCENE_BASELINE_OBJ_INDEX].at(get_node_index("Cube_1", &scene_nodes[SCENE_BASELINE_OBJ_INDEX])).node->get_transform().get_translation();
```
You could store the reference to the transform here, and then set the translation without searching for the SceneNode again.
Like this:
```
	static auto &transform = std::find_if(scene_nodes[SCENE_BASELINE_OBJ_INDEX].begin(),
	                                      scene_nodes[SCENE_BASELINE_OBJ_INDEX].end(),
	                                      [](SceneNode const &scene_node) { return scene_node.node->get_name() == "Cube_1"; })
	                             ->node->get_transform();
	static auto  translation = transform.get_translation();

```

Szymon-Rajczakowski replied:
Change applied. Based on this change, I removed get_node_index(), because this is the only place where this method was called.

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1096 @@
+		time_pass = 0;
+
+		/* Write new position to object */
+		for (uint32_t i = 0; i < scene_nodes[SCENE_BASELINE_OBJ_INDEX].size(); i++)
```
With the above change, you could just set
`		transform.set_translation(translation);
`
and skip the loop here.

Szymon-Rajczakowski replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.h:
```diff
@@ -0,0 +1,169 @@
+		glm::mat4 projection;
+		glm::mat4 background_modelview;
+
+	} ubo_background;
```
The three UBO structs have the first two members in common.
Maybe introduce a new UBO with just those two members to be used in each case, and a special UBO for the baseline and the tessellation part?

Karina-Murawko-Wisniewska-Mobica replied:
UBO refactored, common uniform buffer introduced (used in all 3 pipelines), shaders for tessallation are now simplified, thanks for the suggestion, changes merged

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.h:
```diff
@@ -0,0 +1,169 @@
+		vkb::sg::Node    *node;
+		vkb::sg::SubMesh *sub_mesh;
+	};
+	std::vector<std::vector<SceneNode>> scene_nodes;
```
You never need the the fact that your three scenes are in a vector.
Maybe just having three separate `std::vector<SceneNode>`s would be easier to read?

Szymon-Rajczakowski replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.h:
```diff
@@ -0,0 +1,169 @@
+	void create_descriptor_pool();
+	void setup_descriptor_set_layout();
+	void create_descriptor_sets();
+	int  get_node_index(std::string name, std::vector<SceneNode> *scene_node);
```
Maybe better:
`	int  get_node_index(std::string const & name, std::vector<SceneNode> const & scene_node) const;
`

Szymon-Rajczakowski replied:
get_node_index has been removed so fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.h:
```diff
@@ -0,0 +1,169 @@
+	int  get_node_index(std::string name, std::vector<SceneNode> *scene_node);
+	void selection_indicator(const vkb::sg::PBRMaterial *original_mat, vkb::sg::PBRMaterial *new_mat);
+	void scene_pipeline_divide(std::vector<std::vector<SceneNode>> *scene_node);
+	void draw_from_scene(VkCommandBuffer command_buffer, std::vector<std::vector<SceneNode>> *scene_node, sceneObjType_t scene_index);
```
Maybe a `const & scene_node` is better suited than a `* scene_node`.

Szymon-Rajczakowski replied:
fixed

asuessenbach commented:
Besides the other comments: I think, having lengthy code snippets in the readme.md, without detailed descriptions or comments on the separate steps therein, doesn't help the user too much.
And finally: showcasing the tessellation related dynamic states would probably better when it's compared to some non-dynamic usage. But that might be worth a separate sample, maybe in the performance section.

Karina-Murawko-Wisniewska-Mobica committed:
README simple fixes after code review (#3)

Commit:
Call load_scene
Remove scene member from extended_dynamic_state2 sample

Commit:
Remove unnecessary comment

Szymon-Rajczakowski committed:
Merge pull request #4 from Mateusz-Kowalewski-Mobica/szymon/fixes

Szymon/fixes

Zbigniew-Jedlinski-Mobica committed:
Removal unused variables from extended_dynamic_state2

variables range and depth_range have been removed from extended_dynamic_state

Karina-Murawko-Wisniewska-Mobica committed:
Review fixes in selection_indicator method (#5)

Zbigniew-Jedlinski-Mobica committed:
Merge pull request #6 from zbigniew-jedlinski-mobica/master

Removal unused variables from extended_dynamic_state2

SaschaWillems commented on samples/README.md:
````diff
@@ -268,6 +268,12 @@ Demonstrate how to do conditional rendering, dynamically discarding rendering co
+### [Extended dynamic state 2](./extensions/extended_dynamic_state2)
+**Extension**: [```VK_EXT_extended_dynamic_state2```](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_extended_dynamic_state2.html)
+<br/>
+Demonstrate how to use depth bias, primitive restart, rasterizer discard and patch control points dynamically, which can reduce the number of pipeline objects that are need to be created.
````
Not a native speaker myself, but I think "that are need**ed** to be created" would be the correct spelling here.

Karina-Murawko-Wisniewska-Mobica replied:
Fixed

SaschaWillems commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1065 @@
+	clear_values[0].color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+	clear_values[1].depthStencil = {0.0f, 0};
+
+	int i = -1;
```
Any reason for not just using a loop over i instead of the auto loop like the other samples do?

Karina-Murawko-Wisniewska-Mobica replied:
I already discussed with author and change is required here.
Range-based loop was used to access element in vector, 'i' variable is only needed for framebuffers[i].
But in loop body instead of draw_cmd_buffer the element of vector is accessed by draw_cmd_buffers[i] -  this needs to be fixed.

Karina-Murawko-Wisniewska-Mobica replied:
Changes merged

SaschaWillems commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1065 @@
+		ImGui::ListBox("", &gui_settings.selected_obj, obj_names.data(), obj_cnt);
+		ImGui::NextColumn();
+		if (drawer.checkbox("Depth Bias Enable", &gui_settings.objects[gui_settings.selected_obj].depth_bias))
+		{
```
You can remove this if-statement.

Zbigniew-Jedlinski-Mobica replied:
if-statement was removed

SaschaWillems commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1065 @@
+		if (drawer.checkbox("Depth Bias Enable", &gui_settings.objects[gui_settings.selected_obj].depth_bias))
+		{
+		}
+		if (drawer.checkbox("Rasterizer Discard", &gui_settings.objects[gui_settings.selected_obj].rasterizer_discard))
```
Same here, the if statement is unnecessary.

Zbigniew-Jedlinski-Mobica replied:
if-statement was removed

SaschaWillems commented:
Thank you very much for this new sample. I did a small code review and added a few comments.

The sample works fine on my RTX 2060, but the selection stuff doesn't work properly. I can see that something changes if an object is selected by some kind of color fade, but then the object just starts to flicker at high frequency in a darker color. I guess that part of the code still needs some fixing. Looks like it's not doing an frame time based calculations.

Karina-Murawko-Wisniewska-Mobica committed:
Use reference variable from range-based loop (#7)

MarcinKlosowskiMobica committed:
corrected get_node_index() and update_uniform_buffers() functions after the sample review

Szymon-Rajczakowski committed:
Return color with changed alpha value and skip new_mat (#8)

* Change name of selection_indicator method to get_changed_alpha
get_changed_alpha returns base_color_factor of original_mat with changed alpha value
Remove previous_material (unused variable)
Update README

* Update extended_dynamic_state2.cpp

Fix spaces

* Update extended_dynamic_state2.cpp

Fix spaces

* Fix problem with flicker at high frequency in a darker color

* Fix clang format

MarcinKlosowskiMobica committed:
remove redundant loop

Karina-Murawko-Wisniewska-Mobica committed:
Tessellation related GUI update (#10)

Zbigniew-Jedlinski-Mobica committed:
Removed unnecessary if statement

Removed empty code blocks of if statement in extended_dynamic_state2.cpp

Karina-Murawko-Wisniewska-Mobica committed:
Update README.md

Szymon-Rajczakowski committed:
- Simplify vector of vectors (#11)

* - Simplify vector of vectors
- Add const qualifier

* - Store the reference to the transform and
set the translation without searching for the SceneNode
- Skip the loop and remove unused method

* Change scene_node.at(i) to scene_node[i]

MarcinKlosowskiMobica committed:
corrected V-35 issue

MarcinKlosowskiMobica committed:
added empty lines at the end of files

Zbigniew-Jedlinski-Mobica committed:
Updated UI menu in extended_dynamic_state2

* Changed list_box to combo_box for object names
* Changed two columns to one column in Model section
- Removed width of column

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1021 @@
+	if (drawer.header("Models"))
+	{
+		drawer.checkbox("Selection effect active", &gui_settings.selection_active);
+		int                      obj_cnt = scene_elements_baseline.size();
```
compiler warning on conversion from `size_t` to `int` -> use a `static_cast<int>` or change `obj_cnt` to a a `size_t` (and then the loop variable `i` below as well).

Karina-Murawko-Wisniewska-Mobica replied:
fixed, static_cast used

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1021 @@
+	constexpr float tick_limit = 0.05;
+	constexpr float delta      = 0.05;
+	constexpr float move_step  = 0.0005;
+	static float    time_pass  = 0;
```
three compiler warnings on truncation from double to float -> add 'f' to the constants (0.05 -> 0.05f, ...)

Karina-Murawko-Wisniewska-Mobica replied:
fixed

Karina-Murawko-Wisniewska-Mobica committed:
Add common uniform buffer + minor fixes (#14)

MarcinKlosowskiMobica committed:
corrected coding style issues

MarcinKlosowskiMobica committed:
corrected updates

MarcinKlosowskiMobica committed:
Merge pull request #15 from Mateusz-Kowalewski-Mobica/mklosowski_google_style

corrected coding style issues

Zbigniew-Jedlinski-Mobica committed:
Changing the way of fetching scene elements

Scene elements are fetched as they traverse the scene tree

Mateusz-Kowalewski-Mobica committed:
Merge branch 'KhronosGroup:main' into master

Zbigniew-Jedlinski-Mobica committed:
Update extended_dynamic_state2.cpp

Zbigniew-Jedlinski-Mobica committed:
Update extended_dynamic_state2.cpp

Zbigniew-Jedlinski-Mobica committed:
Merge pull request #16 from Zbigniew-Jedlinski-Mobica/master

Changing the way of fetching scene elements

MarcinKlosowskiMobica committed:
clang format issues fix (#17)

* clang format issues fix

* branch cleanup

MarcinKlosowskiMobica committed:
added information to readme file (#18)

MarcinKlosowskiMobica committed:
changed .at(i) to [i] (#19)

* changed .at(i) to [i]

* corrected clang issues

Karina-Murawko-Wisniewska-Mobica committed:
Refactor README: shorten and comment code snippet (#20)

Karina-Murawko-Wisniewska-Mobica commented:
> Besides the other comments: I think, having lengthy code snippets in the readme.md, without detailed descriptions or comments on the separate steps therein, doesn't help the user too much. 

@asuessenbach 
Code snippets are shortened and comments in them are added to make it more clear.

> I can see that something changes if an object is selected by some kind of color fade, but then the object just starts to flicker at high frequency in a darker color.

@SaschaWillems 
This bug was fixed in one of CR fixes. Other CR comments were addressed.

Karina-Murawko-Wisniewska-Mobica committed:
Merge branch 'KhronosGroup:main' into master

Mateusz-Kowalewski-Mobica committed:
Merge branch 'KhronosGroup:main' into master

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,259 @@
+
+This sample demonstrates how to use `VK_EXT_extended_dynamic_state2` extension, which eliminates the need to create multiple pipelines in case of specific different parameters.
+
+This extension changes how `Depth Bias`, `Primitive Restart`, `Rasterizer Discard` and `Patch Control Points` are managed. Instead of static description during pipeline creation, this extension allows developers to change those parameters by using a function before every draw.
```
Instead of marking plain text as code (`Depth Bias`,...) I would either not mark it, or use the actual code (`VK_DYNAMIC_STATE_DEPTH_BIAS_ENABLE_EXT`,...)

Karina-Murawko-Wisniewska-Mobica replied:
changed to bold

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,259 @@
+| dynamic_state = {} | dynamic_state = {VK_DYNAMIC_STATE_VIEWPORT,<br>VK_DYNAMIC_STATE_SCISSOR,<br>VK_DYNAMIC_STATE_DEPTH_BIAS_ENABLE_EXT,<br>VK_DYNAMIC_STATE_PRIMITIVE_TOPOLOGY_EXT,<br>VK_DYNAMIC_STATE_RASTERIZER_DISCARD_ENABLE_EXT,<br>VK_DYNAMIC_STATE_PRIMITIVE_TOPOLOGY_EXT,<br>VK_DYNAMIC_STATE_PRIMITIVE_RESTART_ENABLE_EXT} |
+| vkCreateGraphicsPipelines(pipeline1)<br>vkCreateGraphicsPipelines(pipeline2)<br>vkCreateGraphicsPipelines(pipeline3)<br>vkCreateGraphicsPipelines(pipeline4) | vkCreateGraphicsPipelines(pipeline1)<br>vkCreateGraphicsPipelines(pipeline2) |
+| draw(model1, pipeline1)<br>draw(model2, pipeline2)<br>draw(model3, pipeline3)<br>draw(model4, pipeline4) | vkCmdSetPrimitiveRestartEnableEXT(model1, primitiveBoolParam)<br>vkCmdSetDepthBiasEnableEXT(model3, depthBiasBoolParam)<br>vkCmdSetRasterizerDiscardEnableEXT(model1,rasterizerBoolParam)<br>vkCmdSetPrimitiveTopologyEXT(model1, VK_PRIMITIVE_TOPOLOGY_TRIANGLE_LIST)<br>draw(model1, pipeline1)< ...
+	
```
Instead of `model1`,..., something like `commandBuffer1`,... might be some better names here.
And the first call to `vkCmdSetDepthBiasEnableEXT` should probably get `model1` as first argument?

Karina-Murawko-Wisniewska-Mobica replied:
changed all modelX to commandBufferX, fixed the mistakes in numbering

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,259 @@
+More details are provided in the sections that follow.
+
+## Pipelines
+Previously developers had to create multiple pipelines for different parameters in `Depth Bias`, `Primitive Restart`, `Rasterizer Discard` and `Patch Control Points`. This is illustrated in static/non-dynamic pipeline creation.
```
As above comment on marking plaint text as code.

Karina-Murawko-Wisniewska-Mobica replied:
removed marking

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
````diff
@@ -0,0 +1,259 @@
+VK_CHECK(vkCreateGraphicsPipelines(get_device().get_handle(), pipeline_cache, 1, &graphics_create, VK_NULL_HANDLE, &pipeline4));
+```
+
+In the above approach if developer would like to change the patch control points number, then for each different number a new pipeline would be required.
````
Maybe also mention the combinatorial explosion of pipelines with different combinations of settings?

Karina-Murawko-Wisniewska-Mobica replied:
We still have issue with presenting dynamic change of the patch control points number.
When we use triangle based model (together with suitable tessellation shaders) change of the  patch control points number is _breaking_ the model.
We are still looking into it.

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,259 @@
+
+In the above approach if developer would like to change the patch control points number, then for each different number a new pipeline would be required.
+
+However, with 'VK_EXT_extended_dynamic_state2' the number of pipelines can be reduced by the possibility to change parameters of `Depth Bias`, `Primitive Restart`, `Rasterizer Discard` and `Patch Control Points` by calling `vkCmdSetDepthBiasEnableEXT`, `vkCmdSetPrimitiveRestartEnableEXT`, `vkCmdSetRasterizerDiscardEnableEXT` and `vkCmdSetPatchControlPointsEXT` respectively before calling `draw_mo ...
```
VK_EXT_extended_dynamic_state2: ' -> `
Again marked plain text as code.

Karina-Murawko-Wisniewska-Mobica replied:
fixed wrong marking, the others removed

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,259 @@
+However, with 'VK_EXT_extended_dynamic_state2' the number of pipelines can be reduced by the possibility to change parameters of `Depth Bias`, `Primitive Restart`, `Rasterizer Discard` and `Patch Control Points` by calling `vkCmdSetDepthBiasEnableEXT`, `vkCmdSetPrimitiveRestartEnableEXT`, `vkCmdSetRasterizerDiscardEnableEXT` and `vkCmdSetPatchControlPointsEXT` respectively before calling `draw_mo ...
+
+With the usage of above functions we can reduce the number of pipelines. Required dynamic states must be enabled and passed to the`VkGraphicsPipelineCreateInfo` struct. 
+
```
Missing space: the `VkGr...`

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,259 @@
+
+With the usage of above functions we can reduce the number of pipelines. Required dynamic states must be enabled and passed to the`VkGraphicsPipelineCreateInfo` struct. 
+
+`VK_DYNAMIC_STATE_PRIMITIVE_TOPOLOGY_EXT` specifies that the topology state in the `VkPipelineInputAssemblyStateCreateInfo` struct only specifies the topology class and the specific topology order and adjacency must be set dynamically with `vkCmdSetPrimitiveTopology` before any drawing commands.
```
Maybe : ...the topology class and the specific... -> ...the topology class. The specific...

Karina-Murawko-Wisniewska-Mobica replied:
fixed

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
```diff
@@ -0,0 +1,259 @@
+		                nullptr);
+vkCmdBindPipeline(draw_cmd_buffer, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline.background);
+
+/* Drawing background */
```
Is primitve topology set inside `draw_model`?

Karina-Murawko-Wisniewska-Mobica replied:
it was missing in code (but there was no visible issues).

added in both

asuessenbach commented on samples/extensions/extended_dynamic_state2/README.md:
````diff
@@ -0,0 +1,259 @@
+add_device_extension(VK_EXT_EXTENDED_DYNAMIC_STATE_2_EXTENSION_NAME);
+```
+
+Additional features are provided by the VkPhysicalDeviceExtendedDynamicState2FeaturesEXT struct:
````
So what?
Does it tell you what's supported or do you use it to enable support??
Maybe just copy the relevant part from the spec here:
If the VkPhysicalDeviceExtendedDynamicState2FeaturesEXT structure is included in the pNext chain of the [VkPhysicalDeviceFeatures2](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPhysicalDeviceFeatures2.html) structure passed to [vkGetPhysicalDeviceFeatures2](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetPhysicalDeviceFeatures2.html), it is filled in to indicate whether each corresponding feature is supported. VkPhysicalDeviceExtendedDynamicState2FeaturesEXT can also be used in the pNext chain of [VkDeviceCreateInfo](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkDeviceCreateInfo.html) to selectively enable these features.

Karina-Murawko-Wisniewska-Mobica replied:
Fixed as suggested (also struct removed, link to specs added)

asuessenbach commented on samples/extensions/extended_dynamic_state2/extended_dynamic_state2.cpp:
```diff
@@ -0,0 +1,1152 @@
+ * 	@fn void ExtendedDynamicState2::create_pipeline()
+ * 	@brief Creating graphical pipelines: baseline, background, tessellation.
+ */
+void ExtendedDynamicState2::create_pipeline()
```
Should probably named `create_pipelines`

Karina-Murawko-Wisniewska-Mobica replied:
Method renamed

Karina-Murawko-Wisniewska-Mobica committed:
README refactor fixes to review comments (#21)

SaschaWillems reviewed (dismissed):
I'm fine with the current state of the sample. It works as expected and is a nice addition.

Karina-Murawko-Wisniewska-Mobica committed:
Fix validation layers errors (#22)

gary-sweet commented:
I can't test this unfortunately since my device doesn't support tessellation. It might have been nice for just that part of the sample to be disabled rather than the entire thing, but it's probably a bit late to change that now.

Karina-Murawko-Wisniewska-Mobica committed:
Merge branch 'KhronosGroup:main' into master

Karina-Murawko-Wisniewska-Mobica commented:
@asuessenbach  Do you have more comments to the PR?

Karina-Murawko-Wisniewska-Mobica committed:
Merge branch 'main' into master

marty-johnson59 commented:
Per discussion last week, we were just waiting for Andreas' review/approval to merge.  Merging..
